djautotask/sync_rest.py

```diff
@@ -0,0 +1,367 @@
+                    self.model_class.__bases__[0].__name__,
+                    instance_id
+                )
+            )
+
+    def update_or_create_instance(self, api_instance):
+        """
+        Creates and returns an instance if it does not already exist.
+        """
+        result = None
+        api_instance = self.remove_null_characters(api_instance)
+        try:
+            instance_pk = api_instance[self.lookup_key]
+            instance = self.model_class.objects.get(pk=instance_pk)
+        except self.model_class.DoesNotExist:
+            instance = self.model_class()
+            result = CREATED
+
+        try:
+            self._assign_field_data(instance, api_instance)
+
+            # This will return the created instance, the updated instance, or
+            # if the instance is skipped an unsaved copy of the instance.
+            if result == CREATED:
+                if self.model_class is models.TicketTracker:
+                    instance.save(force_insert=True)
+                else:
+                    instance.save()
+            elif instance.tracker.changed():
+                instance.save()
+                result = UPDATED
+            else:
+                result = SKIPPED
+        except IntegrityError as e:
+            # This can happen when multiple threads are creating the
+            # same ticket at once.
+            msg = "IntegrityError while attempting to create {}." \
+                  " Error: {}".format(self.model_class, e)
+            logger.error(msg)
+            raise InvalidObjectException(msg)
+
+        if result == CREATED:
+            result_log = 'Created'
+        elif result == UPDATED:
+            result_log = 'Updated'
+        else:
+            result_log = 'Skipped'
+
+        logger.info('{}: {} {}'.format(
+            result_log,
+            self.model_class.__bases__[0].__name__,
+            instance
+        ))
+
+        return instance, result
+
+    def prune_stale_records(self, initial_ids, synced_ids):
+        """
+        Delete records that existed when sync started but were
+        not seen as we iterated through all records from REST API.
+        """
+        stale_ids = initial_ids - synced_ids
+        deleted_count = 0
+        if stale_ids:
+            delete_qset = self.get_delete_qset(stale_ids)
+            deleted_count = delete_qset.count()
+
+            pre_delete_result = None
+            if self.pre_delete_callback:
+                pre_delete_result = self.pre_delete_callback(
+                    *self.pre_delete_args
+                )
+            logger.info(
+                'Removing {} stale records for model: {}'.format(
+                    len(stale_ids), self.model_class.__bases__[0].__name__,
+                )
+            )
+            delete_qset.delete()
+            if self.post_delete_callback:
+                self.post_delete_callback(pre_delete_result)
```

Comment:
Remove code here related to `pre_delete_result`, `pre_delete_callback`, and `post_delete_callback`